ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java

```diff
@@ -373,9 +390,23 @@ private SensitivityComputer getSensitivityComputerForOptimizationBasedOn(FlowRes
         }
     }
 
-    private synchronized void updateOptimalLeaf(Leaf leaf) {
+    private synchronized void updateOptimalLeaf(Leaf leaf, NetworkActionCombination networkActionCombination) {
         if (improvedEnough(leaf)) {
-            optimalLeaf = leaf;
+            // nominal case: stop criterion hasn't been reached yet
+            if (combinationFulfillingStopCriterion.isEmpty() && leaf.getCost() < optimalLeaf.getCost()) {
+                optimalLeaf = leaf;
+                if (stopCriterionReached(leaf)) {
+                    topLevelLogger.info("Stop criterion reached, other threads may skip optimization.");
+                    combinationFulfillingStopCriterion = Optional.of(networkActionCombination);
+                }
+            }
+            // special case: stop criterion has been reached
+            if (combinationFulfillingStopCriterion.isPresent()
+                && stopCriterionReached(leaf)
+                && orderNetworkActionCombinationsRandomly(networkActionCombination, combinationFulfillingStopCriterion.get()) < 0) {
+                optimalLeaf = leaf;
+                combinationFulfillingStopCriterion = Optional.of(networkActionCombination);
+            }
```

Comment:
we only go in there `if (improvedEnough(leaf))`

Souldn't we put this outside of this if ?

Reply:
ok, i just see you changed improvedEnough, nevermind. :p

Reply:
yeah I had to change the logic slightly to be able to check we respect min impacts, and still be able to update the optimalLeaf with a leaf with a higher cost that still reaches the stop criterion